src/store/data_layer/data_store.js

```diff
@@ -0,0 +1,207 @@
+      // At this point, we have an array of all intervals where we have data stored
+      // We have the start and end times for each interval (inclusive)
+      // Now, we need to determine if there are any gaps.
+      let returnArr = [] // An array of interval pairs, ie: [[start, end]]
+      if (timestamps[0] > payload.start) {
+        // The first gap is between the start and the first timestamp
+        returnArr.push([payload.start, timestamps[0]])
+      }
+      // Iterate over the remaining timestamps and search for gaps
+      for (let index = 1; index < timestamps.length; index += 2) {
+        if (timestamps[index + 1]) {
+          returnArr.push([timestamps[index], timestamps[index + 1]])
+        } else if (timestamps[index] < payload.end) {
+          returnArr.push([timestamps[index], payload.end])
+        }
+      }
+      // Return an array of interval pairs
+      return returnArr
+    }
+  },
+
+  // Retrieves data from the cache if it exists
+  // If the data does not exist, it:
+  //   1. Requests for the data from the API
+  //   2. Caches the datums
+  //   3. Retrieves the data from the cache
+  // Parameters:
+  //  meterId: integer that uniquely identifies the meter
+  //  start: integer representing linux epoch time of the start of the required interval
+  //  end: integer representing linux epoch time of the end of the required interval
+  //  uom: the unit of measure/metering point to request data for
+  //  classInt: An integer that corresponds to the type of meter we are reading from
+  async getData (store, payload) {
+    // Does the cache contain the data?
+    const missingIntervals = await this.dispatch('dataStore/findMissingIntervals', {
+      meterId: payload.meterId,
+      start: payload.start,
+      end: payload.end,
+      uom: payload.uom
+    })
+    if (missingIntervals.length > 0) {
+      // The cache does not contain all of the data
+      // Add it to the cache
+      let promises = []
+      missingIntervals.forEach(async (interval) => { // For each interval, request for the data
+        if (interval.length === 0) interval.push(Math.round(new Date().getTime() / 1000))
+        promises.push(API.data(payload.meterId, interval[0], interval[1], payload.uom, payload.classInt))
+      })
+
+      // Save all of the new data to the cache
+      await Promise.all(promises).then(responses => {
+        // The data looks like an array of these objects:
+        // {
+        //   accumulated_real: -13385083 // This key can change based on the uom
+        //   id: 4596804
+        //   time: 1597284900
+        // }
+        responses.forEach(datumArray => {
+          datumArray.forEach(datum => {
+            this.commit('dataStore/addToCache', {
+              datetime: datum.time,
+              meterId: payload.meterId,
+              uom: payload.uom,
+              value: datum[payload.uom]
+            })
+          })
+        })
+      })
+    } else {
+      console.log('')
```

Comment:
Debug?

Reply:
You know it.

Reply:
> If this module were to morph into a complete data broker and cacher with a formalized request service that would be pretty cool. That may also be a little overkill, and require moving lots of preexisting code.
> 

That was the idea. I'd like to create classes/models for the different data types on the frontend, and have one central cache shared by all of the meters. In the future, I'd like to be able to limit the total cache size based on the type of device, too.

It would also be cool to cache the results of various calculations (such as the regression used to determine trends). I feel that abstracting this into a single data storage layer would be beneficial.

Reply:
Im strongly suggesting my idea below just cause I think it sounds cool....

Reply:
And you know minimizing the code base. Cause it is kind of large for student projects with high turnover. To me it just seems like the classes/models in the data store are more less the same as the modules in the vuex store.